### ISSUE TITLE
`scipy.signal.{medfilt,medfilt2d}` hit "Windows fatal exception: access violation" with SciPy 1.7.0 on Windows

### ISSUE THREAD

**Issue by leofang** [scipy/scipy#14273]
Since yesterday with the new SciPy 1.7.0 release, CuPy's Windows CI is experiencing the following fatal error (cupy/cupy#5364) and the CI was interrupted.

#### Reproducing code example:
The failing tests can be found here:
https://github.com/cupy/cupy/blob/4469fae998df33c72ff40ef954cb08b8f0004b18/tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py#L250-L281

The failure is reproducible only on Windows and **not seen on Linux**.

#### Error message:
```
01:07:54.763807 STDERR 1188] | Windows fatal exception: access violation | &nbsp;
01:07:54.763807 STDERR 1188] | &nbsp; | &nbsp;
01:07:54.763807 STDERR 1188] | Current thread 0x0000121c (most recent call first): | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\scipy\signal\signaltools.py", line 1538 in medfilt | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\scipy\signal\signaltools.py", line 1851 in medfilt2d | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Windows\TEMP\flexci\run-00151477\work\src\tests\cupyx_tests\scipy_tests\signal_tests\test_signaltools.py", line 281 in test_medfilt2d | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\cupy\testing\_helper.py", line 47 in _call_func | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\cupy\testing\_helper.py", line 90 in _call_func_numpy | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\cupy\testing\_helper.py", line 101 in _call_func_numpy_cupy | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\cupy\testing\_helper.py", line 265 in test_func | &nbsp;
01:07:54.763807 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\cupy\testing\_helper.py", line 825 in test_func | &nbsp;
01:07:54.764804 STDERR 1188] | File "C:\Development\Python\Python37\lib\unittest\case.py", line 615 in run | &nbsp;
01:07:54.764804 STDERR 1188] | File "C:\Development\Python\Python37\lib\unittest\case.py", line 663 in __call__ | &nbsp;
01:07:54.764804 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\unittest.py", line 321 in runtest | &nbsp;
01:07:54.764804 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 162 in pytest_runtest_call | &nbsp;
01:07:54.765777 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\callers.py", line 187 in _multicall | &nbsp;
01:07:54.765777 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 87 in &lt;lambda&gt; | &nbsp;
01:07:54.765777 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 93 in _hookexec | &nbsp;
01:07:54.765777 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\hooks.py", line 286 in __call__ | &nbsp;
01:07:54.765777 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 255 in &lt;lambda&gt; | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 311 in from_call | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 255 in call_runtest_hook | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 215 in call_and_report | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 126 in runtestprotocol | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\runner.py", line 109 in pytest_runtest_protocol | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\callers.py", line 187 in _multicall | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 87 in &lt;lambda&gt; | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 93 in _hookexec | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\hooks.py", line 286 in __call__ | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\main.py", line 348 in pytest_runtestloop | &nbsp;
01:07:54.769663 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\callers.py", line 187 in _multicall | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 87 in &lt;lambda&gt; | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 93 in _hookexec | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\hooks.py", line 286 in __call__ | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\main.py", line 323 in _main | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\main.py", line 269 in wrap_session | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\main.py", line 316 in pytest_cmdline_main | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\callers.py", line 187 in _multicall | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 87 in &lt;lambda&gt; | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\manager.py", line 93 in _hookexec | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pluggy\hooks.py", line 286 in __call__ | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\config\__init__.py", line 163 in main | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\_pytest\config\__init__.py", line 185 in console_main | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\site-packages\pytest\__main__.py", line 5 in &lt;module&gt; | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\runpy.py", line 85 in _run_code | &nbsp;
01:07:54.781377 STDERR 1188] | File "C:\Development\Python\Python37\lib\runpy.py", line 193 in _run_module_as_main | 12s
01:08:06.784203 STDOUT 1188] | Last 10 lines from the test output: | &nbsp;
01:08:06.812846 STDOUT 1188] | ........................................................................ [ 75%] | &nbsp;
01:08:06.814324 STDOUT 1188] | ........................................................................ [ 75%] | &nbsp;
01:08:06.814324 STDOUT 1188] | ........................................................................ [ 75%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 75%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 75%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 76%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 76%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 76%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ........................................................................ [ 76%] | &nbsp;
01:08:06.814604 STDOUT 1188] | ............FFFF....FFFF
```

#### Scipy/Numpy/Python version information:
Python: 3.7.2
NumPy: 1.20.3
SciPy: 1.7.0

cc: @kmaehashi @emcastillo

**Comment by tylerjereddy** [comment]
Any chance it is related to the release of the GIL recently added: https://github.com/scipy/scipy/pull/13532

**Comment by emcastillo** [comment]
I hadn't verified this yet in a windows machine, but the CuPy tests fail with a code like this:

```python
import numpy
import scipy.signal

sizes = [(10,), (5, 10), (10, 5), (5, 6, 10)]
kernel_sizes = [3, (3, 3, 5)]
dtypes = [numpy.uint8, numpy.float32, numpy.float64]

for size in sizes:
    for kernel_size in kernel_sizes:
        for dtype in dtypes:
            volume = numpy.arange(numpy.prod(size), dtype=dtype).reshape(size)
            if isinstance(kernel_size, tuple):
                kernel_size = kernel_size[:volume.ndim]
            scipy.signal.medfilt(volume, kernel_size)
```

**Comment by ilayn** [comment]
I can't reproduce these on Windows 10, Python 3.9.2, NumPy 1.20.1, Scipy 1.17.0, all pip installed.

**Comment by rgommers** [comment]
Just to make sure, since I cannot access the CI logs - this used a SciPy wheel from PyPI? Or a conda-forge package or from-source build? And is it crashing only on Python 3.7, or also 3.8/3.9?

**Comment by peterbell10** [comment]
Tried with python 3.7 and scipy installed through pip, conda defaults and conda-forge. None of them reported any failure running the reproducer. 

FWIW, I suspect this might be related to #13262. It's the only change that effects both `medfilt` and `medfilt2d`. It also exposes new code, since inputs aren't always promoted to double any more.

**Comment by leofang** [comment]
Thanks @ilayn @peterbell10 for helping test it!

IIUC it's using a SciPy wheel from PyPI, but I don't know about how Python is installed (@kmaehashi or @emcastillo should know better). I think this CI sticks to a particular combination of software versions, so it's not testing a large build matrix. I don't have access to my Windows machine for the moment. Once I regain access I could give it a try.

btw the CI is at https://ci.preferred.jp/cupy.experimental.win.cuda100/. PFN puts it behind a Google login, but with a valid Google account anyone can access the logs.

**Comment by rgommers** [comment]
A bit of a puzzle why it's only reproducible on that CI setup. Maybe caused by some packaging thing, like the wheel bundling the `msvcp140.dll` MSVC C++ runtime library, and that setup using a different version of it.

> btw the CI is at https://ci.preferred.jp/cupy.experimental.win.cuda100/. PFN puts it behind a Google login, but with a valid Google account anyone can access the logs.

Thanks, I can access that now. There's nothing to see anymore - the logs with scipy 1.7.0 have disappeared (too long ago), and now scipy is pinned to `<1.7.0`. I guess there won't be more to see than in the traceback above though.

Nice function name by the way:) - `RunOrDie python -m pip install Cython 'scipy<1.7' optuna`

**Comment by emcastillo** [comment]
I will try to isolate the snippet above in our CI, sorry for the delay!

**Comment by leofang** [comment]
> Thanks, I can access that now. There's nothing to see anymore - the logs with scipy 1.7.0 have disappeared (too long ago), and now scipy is pinned to `<1.7.0`. I guess there won't be more to see than in the traceback above though.

hmmm I didn't realize the history list would be truncated aggressively. Anyway the failing CI can still be accessed, ex: https://ci.preferred.jp/cupy.experimental.win.cuda100/73846/. It's just not shown on the recent history anymore.

**Comment by kmaehashi** [comment]
I've made a minimum reproducer.

```
import numpy, scipy.signal
arr = numpy.arange(50, dtype=numpy.float16).reshape((5, 10))
scipy.signal.medfilt2d(arr, 3)
```

In SciPy 1.6.3 the was correctly raised when float16 is given.

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Development\Python\Python37\lib\site-packages\scipy\signal\signaltools.py", line 1839, in medfilt2d
    return sigtools._medfilt2d(image, kernel_size)
ValueError: 2D median filter only supports uint8, float32, and float64.
```

However, the above code causes a hard crash in SciPy 1.7.0. Reproducible Python 3.7.2 / 3.9.0 on Windows, both using PyPI wheel.

**Comment by rgommers** [comment]
We don't build Windows nightlies; the fix for this should land in 1.7.1 in some weeks. Timing TBD, we hope to also resolve the macOS arm64 issues and provide wheels. 

Given that this issue only happens for code that was raising an exception before, I think it shouldn't affect too many users, so that timing is okay. Please let us know in case that assessment is incomplete and we should release a fix sooner @leofang @kmaehashi.

**Comment by leofang** [comment]
Thanks, Ralf!
